Project 4 - Predictive Analysis using scikit-learn

Name: Anisul Arman

Course_ID: IS 362 Data Acquisition and Management . 

## Description


### Assignment – Predictive Analysis using scikit-learn

Start with the mushroom data in the pandas DataFrame that you constructed in your “Assignment – Preprocessing Data with sci-kit learn.”
• Use scikit-learn to determine which of the two predictor columns that you selected (odor and one other column of your choice) most accurately predicts whether or not a mushroom is poisonous. There is an additional challenge here—to use scikit-learn’s predictive classifiers, you’ll want to convert each of your two (numeric categorical) predictor columns into a set of columns. See for one approach pandas get_dummies() method.
• Clearly state your conclusions along with any recommendations for further analysis.

In [1]:
# i am isntalling my libaries
!pip install pandas
!pip install requests
!pip install sklearn

  Using cached tzdata-2025.2-py2.py3-none-any.whl (347 kB)
  Using cached pytz-2025.2-py2.py3-none-any.whl (509 kB)


You should consider upgrading via the 'C:\Users\Admin\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


  Using cached requests-2.32.3-py3-none-any.whl (64 kB)
  Using cached idna-3.10-py3-none-any.whl (70 kB)


You should consider upgrading via the 'C:\Users\Admin\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


Using legacy 'setup.py install' for sklearn, since package 'wheel' is not installed.
    Running setup.py install for sklearn: started
    Running setup.py install for sklearn: finished with status 'done'


    ERROR: Command errored out with exit status 1:
     command: 'C:\Users\Admin\AppData\Local\Programs\Python\Python310\python.exe' -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\Admin\\AppData\\Local\\Temp\\pip-install-jvz_xkqx\\sklearn_453798c08ec54640870ecd71da988ba9\\setup.py'"'"'; __file__='"'"'C:\\Users\\Admin\\AppData\\Local\\Temp\\pip-install-jvz_xkqx\\sklearn_453798c08ec54640870ecd71da988ba9\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base 'C:\Users\Admin\AppData\Local\Temp\pip-pip-egg-info-51k_dqi1'
         cwd: C:\Users\Admin\AppData\Local\Temp\pip-install-jvz_xkqx\sklearn_453798c08ec54640870ecd71da988ba9\
    Complete output (15 lines):
    The 'sklearn' PyPI package is deprecated, use 'scikit-learn'
  

In [2]:
import pandas as pd
import requests
import zipfile
import requests
import io
import matplotlib.pyplot as plt


from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score


import warnings
warnings.filterwarnings("ignore")

In [3]:
# Download and unzip the file
url = "https://archive.ics.uci.edu/static/public/73/mushroom.zip"
r = requests.get(url)

with zipfile.ZipFile(io.BytesIO(r.content)) as z:
    z.extractall()


In [4]:
data = "agaricus-lepiota.data"

# load data
df = pd.read_csv(data)
df.head()

p  x  s  n  t p.1  f  c n.1  k  ... s.2  w w.1 p.2 w.2  o p.3 k.1 s.3  u
0  e  x  s  y  t   a  f  c   b  k  ...   s  w   w   p   w  o   p   n   n  g
1  e  b  s  w  t   l  f  c   b  n  ...   s  w   w   p   w  o   p   n   n  m
2  p  x  y  w  t   p  f  c   n  n  ...   s  w   w   p   w  o   p   k   s  u
3  e  x  s  g  f   n  f  w   b  k  ...   s  w   w   p   w  o   e   n   a  g
4  e  x  y  y  t   a  f  c   b  n  ...   s  w   w   p   w  o   p   k   n  g

[5 rows x 23 columns]

In [5]:
# Map the columns name found on the website to the datasets
# Column names from the data dictionary
columns = [
    'class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
    'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color', 'stalk-shape',
    'stalk-root', 'stalk-surface-above-ring', 'stalk-surface-below-ring',
    'stalk-color-above-ring', 'stalk-color-below-ring', 'veil-type', 'veil-color',
    'ring-number', 'ring-type', 'spore-print-color', 'population', 'habitat'
]

# Step 3: Read the CSV into pandas
data = pd.read_csv(data, header=None, names=columns)
data.head()

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0     p         x           s         n       t    p               f   
1     e         x           s         y       t    a               f   
2     e         b           s         w       t    l               f   
3     p         x           y         w       t    p               f   
4     e         x           s         g       f    n               f   

  gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0            c         n          k  ...                        s   
1            c         b          k  ...                        s   
2            c         b          n  ...                        s   
3            c         n          n  ...                        s   
4            w         b          k  ...                        s   

  stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                      w                      w         p          w   
1                      w                      w         p          w   
2                      w                      w         p          w   
3                      w                      w         p          w   
4                      w                      w         p          w   

  ring-number ring-type spore-print-color population habitat  
0           o         p                 k          s       u  
1           o         p                 n          n       g  
2           o         p                 n          n       m  
3           o         p                 k          s       u  
4           o         e                 n          a       g  

[5 rows x 23 columns]

In [6]:
data.columns

Index(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

From my online research, I found that the gill size of mushrooms can be an indicator of their poisonous nature. Mushrooms have gills located underneath their caps, which are responsible for producing and releasing spores. In some cases, poisonous mushrooms tend to have very tight, crowded, or delicate gills compared to edible varieties.


https://pmc.ncbi.nlm.nih.gov/articles/PMC10785791/

In [7]:
# Subset the data
selected_columns = ['class', 'odor', 'gill-size']
df = data[selected_columns]
df.head()

class odor gill-size
0     p    p         n
1     e    a         b
2     e    l         b
3     p    p         n
4     e    n         b

In [8]:
# Add meaningful names
df.columns = ['edibility', 'odor', 'gill_size']
df.head()


edibility odor gill_size
0         p    p         n
1         e    a         b
2         e    l         b
3         p    p         n
4         e    n         b

In [9]:
# Map categorical codes to numbers e=edible (0), p=poisonous (1)
df['edibility'] = df['edibility'].map({'e': 0, 'p': 1})  

# Map odor
odor_mapping = {char: idx for idx, char in enumerate(df['odor'].unique())}
df['odor'] = df['odor'].map(odor_mapping)

# Map gill_size, 'b' = broad, 'n' = narrow
gill_size_mapping = {'b': 0, 'n': 1}
df['gill_size'] = df['gill_size'].map(gill_size_mapping)

df.head()

edibility  odor  gill_size
0          1     0          1
1          0     1          0
2          0     2          0
3          1     0          1
4          0     3          0

In [10]:
# Convert selected columns ('odor' and 'gill_size') using get_dummies
selected_columns = ['odor', 'gill_size']
df_dummies = pd.get_dummies(df[selected_columns])

In [11]:
# Split data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(
    df_dummies, df['edibility'], test_size=0.3, random_state=42
)

In [12]:
# Train Decision Tree Classifier
clf = DecisionTreeClassifier(random_state=42)
clf.fit(X_train, y_train)


DecisionTreeClassifier(random_state=42)

In [13]:
# Predict and evaluate accuracy
y_pred = clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

In [14]:
# Display accuracy and feature importance
print("Accuracy:", accuracy)
feature_importances = pd.Series(clf.feature_importances_, index=df_dummies.columns).sort_values(ascending=False)
print("Feature Importances:\n", feature_importances)

Accuracy: 0.9840032813781788
Feature Importances:
 odor         0.99759
gill_size    0.00241
dtype: float64


### Conclusion and Recommendation:

Upon completion of this analysis, it was clearly observed that the odor attribute had a dominant predictive power (importance score: 0.99759) in determining mushroom edibility, substantially outperforming gill_size, which had a significantly lower importance score of 0.00241. This result strongly suggests that odor alone is nearly sufficient for accurately classifying mushrooms as poisonous or edible.

Given these findings, it is recommended that further analyses incorporate additional mushroom attributes to enhance model accuracy and robustness. Employing advanced validation techniques such as cross-validation and exploring alternative classifiers, like Random Forest or Gradient Boosting, would help validate these findings and potentially uncover other valuable predictive relationships within the dataset.